substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs

```diff
@@ -1456,25 +1462,37 @@ where
 		);
 		let api = self.api.clone();
 
-		// We keep track of everything we prune so that later we won't add
-		// transactions with those hashes from the retracted blocks.
-		let mut pruned_log = HashSet::<ExtrinsicHash<ChainApi>>::new();
+		// Collect extrinsics on the enacted path in a map from block hn -> extrinsics.
+		let mut extrinsics = self.collect_extrinsics(tree_route.enacted()).await;
 
 		// Create a map from enacted blocks' extrinsics to their `provides`
 		// tags based on inactive views.
-		let known_provides_tags = self.collect_provides_tags_from_view_store(tree_route).await;
+		let known_provides_tags = Arc::new(
+			self.collect_provides_tags_from_view_store(
+				tree_route,
+				extrinsics.values().flatten().map(|tx| view.pool.hash_of(tx)).collect(),
+			)
+			.await,
+		);
+
 		debug!(target: LOG_TARGET, "update_view_with_fork: txs to tags map length: {}", known_provides_tags.len());
+
+		// We keep track of everything we prune so that later we won't add
+		// transactions with those hashes from the retracted blocks.
+		let mut pruned_log = HashSet::<ExtrinsicHash<ChainApi>>::new();
 		future::join_all(tree_route.enacted().iter().map(|hn| {
 			let api = api.clone();
 			let known_provides_tags = known_provides_tags.clone();
+			let xts = extrinsics.remove(&hn.hash).unwrap_or_else(|| Vec::new());
```

Comment:
will it work here?
```suggestion
			let xts = extrinsics.remove(&hn.hash).unwrap_or_default();
```

Reply:
That definitely works, thanks: [8985969](https://github.com/paritytech/polkadot-sdk/pull/7980/commits/89859699b583707a1872f834d60e6626ba2a2eb9)